### PR TITLE
REFACTOR: Opinion: TXT records are one long string

### PR DESCRIPTION
**Pull request by tlimoncelli**
This big PR implements an internal change to how TXT records are handled. There should be no user-visible changes.

* TXT-handling code is now simplified with fewer edge-cases
* A new opinion is added ("Opinion # 8 TXT Records are one long string")
* The `get-zone` subcommand now generates TAB-separated files that are more correct for TXT records (no quotes).

TXT record handling was complex because we abstract them at the wrong layer.

OLD: We try to maintain TXT records as a list of sub-strings, creating complexity at the main level, the diff engine, and the provider level.

NEW: TXT records are stored as a single (possibly very long) string. If a provider's API requires long strings to be split up, joined, escaped, or quoted the provider must handle that internally; it should be invisible to the  main code and (more importantly) invisible to the user.

### CODE REVIEW

File: models/dnsrr.go [hunk 1]
```diff
@@ -73,7 +83,17 @@ func RRtoRC(rr dns.RR, origin string) (RecordConfig, error) {
...
 		err = rc.SetTargetTLSA(v.Usage, v.Selector, v.MatchingType, v.Certificate)
 	case *dns.TXT:
-		err = rc.SetTargetTXTs(v.Txt)
+		if fixBug {
+			t := strings.Join(v.Txt, "")
+			te := t
+			te = strings.ReplaceAll(te, `\\`, `\`)
+			te = strings.ReplaceAll(te, `\"`, `"`)
+			//fmt.Fprintf(os.Stderr, "DEBUG: RRtoRC inboundv=%v\n", t)
+			//fmt.Fprintf(os.Stderr, "DEBUG: RRtoRC decodedv=%v\n", te)
```
Comment (tlimoncelli) [inline models/dnsrr.go:92]:
```suggestion
```

File: pkg/diff2/compareconfig.go [hunk 2]
```diff
@@ -103,7 +103,9 @@ func NewCompareConfig(origin string, existing, desired models.Records, compFn Co
 		labelMap: map[string]bool{},
 		keyMap:   map[models.RecordKey]bool{},
 	}
+	//printer.Printf("DEBUG: EXISTING:\n")
```
Comment (tlimoncelli) [inline pkg/diff2/compareconfig.go:106]:
```suggestion
```

File: pkg/diff2/compareconfig.go [hunk 3]
```diff
@@ -103,7 +103,9 @@ func NewCompareConfig(origin string, existing, desired models.Records, compFn Co
...
 	cc.addRecords(existing, true) // Must be called first so that CNAME manipulations happen in the correct order.
+	//printer.Printf("DEBUG: DESIRED:\n")
```
Comment (tlimoncelli) [inline pkg/diff2/compareconfig.go:108]:
```suggestion
```

File: pkg/normalize/flatten.go [hunk 4]
```diff
@@ -32,7 +32,7 @@ func flattenSPFs(cfg *models.DNSConfig) []error {
 		// flatten all spf records that have the "flatten" metadata
 		for _, txt := range txtRecords {
 			var rec *spflib.SPFRecord
-			txtTarget := strings.Join(txt.TxtStrings, "")
+			txtTarget := txt.GetTargetField()
```
Comment (tlimoncelli) [inline pkg/normalize/flatten.go:35]:
```suggestion
			txtTarget := txt.GetTargetTXTJoined()
```

File: pkg/prettyzone/prettyzone.go [hunk 5]
```diff
@@ -138,7 +140,14 @@ func (z *ZoneGenData) generateZoneFileHelper(w io.Writer) error {
...
 		// the remaining line
-		target := rr.GetTargetCombined()
+		var target string
+		if typeStr == "TXT" {
+			target = rr.GetTargetCombinedFunc(txtutil.EncodeQuoted)
+			printer.Printf("DEBUG: pretty txt outbounds=%s\n", rr.GetTargetField())
+			printer.Printf("DEBUG: pretty txt  encodedv=%v\n", target)
+		} else {
+			target = rr.GetTargetCombined()
+		}
```
Comment (tlimoncelli) [inline pkg/prettyzone/prettyzone.go:150]:
```suggestion
		target = rr.GetTargetCombinedFunc(txtutil.EncodeQuoted)
```

**Review by costasd — COMMENTED** [review]
:wave: Changes are needed for NS1 to pass integration and manual tests. See #2644. Thanks!

**Comment by imlonghao** [comment]
`PORKBUN` passed using `go test -v -verbose -provider PORKBUN -start 16 -end 17`

As for `CLOUDNS`, my previous working configuration which has a very long single string DKIM TXT record, failed in v4.7.0

**Comment by tlimoncelli** [comment]
> `PORKBUN` passed using `go test -v -verbose -provider PORKBUN -start 16 -end 17`

Thanks for letting me know!

> As for `CLOUDNS`, my previous working configuration which has a very long single string DKIM TXT record, failed in v4.7.0

Glad to know!  If you'd like help debugging, please open a new issue.  I'd be glad to assist!

Tom

**Pull request merged**